In [4]:
import pandas as pd
from sklearn.datasets import fetch_california_housing

# Load the housing dataset
housing = fetch_california_housing()

In [ ]:
#feature names and basic info
X = pd.DataFrame(housing.data, columns=housing.feature_names) 
y = pd.Series(housing.target, name='med_house_value')

print(X.head(5))
print(X.info())
print(X.describe())

In [ ]:
#train/test split
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# training linear regression model
from sklearn.linear_model import LinearRegression
model = LinearRegression()
model.fit(X_train, y_train)

#making predictions
y_pred = model.predict(X_test)

#evaluating model performance
import numpy as np
from sklearn.metrics import mean_squared_error, r2_score
mse= mean_squared_error(y_test, y_pred)
rmse= np.sqrt(mse)
r2= r2_score(y_test, y_pred)

print('MSE:', mse)
print('RMSE:',rmse)
print('R2:', r2)


MSE: 0.5558915986952446
RMSE: 0.7455813830127765
R2: 0.5757877060324506


## What does the R² score tell us about model performance?
*The R2 score represents the proprotion of variation in median house values that is then explained by the model's features. In this example, it tells us that the model explains 57.6% of the variation in median house values.


In [ ]:
#features with strongest impact
coefficients = pd.DataFrame({
    "Feature": X.columns,
    "Coefficient": model.coef_
}).sort_values(by="Coefficient", key=abs, ascending=False)

coefficients

,Feature,Coefficient
3,AveBedrms,0.783145
0,MedInc,0.448675
7,Longitude,-0.433708
6,Latitude,-0.419792
2,AveRooms,-0.123323
1,HouseAge,0.009724
5,AveOccup,-0.003526
4,Population,-0.000002


## Which features seem to have the strongest impact on predictions based on the model’s coefficients?

*Average Bedrooms, Median Income, and Longitude are the features that are showing the strongest coefficients. 
*Med Income having the strongest impact on the housing pricing predictions suggests that areas with higher median incomes would have higher home values.

## How well do the predicted values match the actual values?

*The predicted values match the actual pretty well with an R2 of .576 the model 

In [ ]:
#choosing 3 features
selected_features= ['AveBedrms', 'MedInc','Longitude']

X_Simple=X[selected_features]
#training the simplified model
X_train_s, X_test_s, y_train_s, y_test_s = train_test_split(X_Simple, y, test_size=0.2, random_state=42)

simple_model = LinearRegression()
simple_model.fit(X_train_s, y_train_s)

y_pred_s=simple_model.predict(X_test_s)

#Testing the simplified model
mse_s= mean_squared_error(y_test_s, y_pred_s)
rmse_s=np.sqrt(mse_s)
r2_s= r2_score(y_test_s, y_pred_s)

print('MSE:', mse_s)
print('RMSE', rmse_s)
print('R2', r2_s)



MSE: 0.7070831927336602
RMSE 0.8408823893587379
R2 0.4604102959651172


*I chose these 3 features because my earlier coefficent testing indicated they they were 3 strongest housing price predictors.

*The simplified model is not as good as the full model across all of the given metrics. The R2 decreses from 0.576 to 0.46 meaning that it explains less variation in housing prices. The RMSE increases from 0.746 to 0.841 which indicates more prediction error per home. This means that by removing the features it reduces the model's predictive power and increases error.

*In practice, you obviously would want to use the full model due to the greater accuracy. However, I could see how the simplified model is easier to interpret and would be easier to create when just looking for ballpark estimates.